src/parseTools.js

```diff
@@ -910,6 +915,57 @@ function to64(x) {
   return `BigInt(${x})`;
 }
 
+// Generates an expression that tests whether a pointer (either a BigInt or a signed int32 JS Number) is aligned to the given byte multiple.
+function isPtrAligned(ptr, alignment) {
+  if (MEMORY64) return /^\d+$/.test(alignment) ? `BigInt(${ptr}) % ${alignment}n == 0` : `${ptr} % BigInt(${alignment}) == 0`;
+  return `${ptr} % ${alignment} == 0`;
+}
+
+const MAX_MEMORY53 = 2 ** 53; // == 9007199254740992
+
+// Converts a pointer (either a BigInt or a signed int32 JS Number) in-place to an index on the heap (a BigInt or an unsigned JS Number).
+// N.B. in ASSERTIONS mode may generate two statements (in form "a;b;").
+function convertPtrToIdx(ptr, accessWidth) {
+  const assertPointerAlignment = ASSERTIONS ? `assert(${isPtrAligned(ptr, accessWidth)});` : '';
+
+  let conversion;
+  if (MEMORY64) {
+    // if our address space would ever get larger than 53 bits, we could do something like this:
+    // if (MAXIMUM_MEMORY > MAX_MEMORY53) conversion = /^\d+$/.test(accessWidth) ? `${ptr} >>= ${accessWidth}n` : `${ptr} >>= BigInt(${accessWidth})`;
+    assert(MAXIMUM_MEMORY <= MAX_MEMORY53);
+    conversion = `${ptr} = Number(BigInt(${ptr}) >> BigInt(${accessWidth}))`;
+  } else if (MAXIMUM_MEMORY > 2 * 1024 * 1024 * 1024) {
+    conversion = `${ptr} >>>= ${accessWidth}`;
+  } else {
+    conversion = accessWidth ? `${ptr} >>= ${accessWidth}` : '';
+  }
+
+  return assertPointerAlignment + conversion;
+}
+
+// Returns a pointer (either a BigInt or a signed int32 JS Number) shifted to an index on the heap (a BigInt or an unsigned JS Number).
+function ptrToIdx(ptr, accessWidth) {
+  if (MEMORY64) {
+    assert(MAXIMUM_MEMORY <= MAX_MEMORY53);
+    return `Number(BigInt(${ptr}) >> BigInt(${accessWidth}))`;
+  }
+  if (MAXIMUM_MEMORY > 2 * 1024 * 1024 * 1024) return `${ptr} >>> ${accessWidth}`;
+  return accessWidth ? `${ptr} >> ${accessWidth}` : ptr;
+}
+
+// Converts a given JS Number (e.g. a pointer offset) to a heap index type (either a BigInt or a JS Number)
+function createHeapIdx(number) {
+  if (MAXIMUM_MEMORY > MAX_MEMORY53) return /^-?\d+$/.test(number) ? `${number}n` : `BigInt(${number})`;
+  return number;
+}
+
+// Browsers might not support passing BigInts to different API entry points that take in addresses, so convert those to Number()s before calling into the browser APIs.
+function idxToMemory53(heapIndex) {
```

Comment:
I would hope this is not needed since all pointers should be converted to int53 numbers when they arrive from native code.

Reply:
Ok, i will give removing this a try